kratos/sources/model_part.cpp

```diff
@@ -13,21 +13,109 @@
 
 // System includes
 #include <sstream>
+#include <type_traits>
 
 // External includes
 
 // Project includes
 #include "includes/define.h"
 #include "includes/model_part.h"
 #include "includes/exception.h"
-#include "utilities/parallel_utilities.h"
+#include "utilities/reduction_utilities.h"
 
 namespace Kratos
 {
 
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, ALL_ENTITIES, 0);
 KRATOS_CREATE_LOCAL_FLAG(ModelPart, OVERWRITE_ENTITIES, 1);
 
+namespace ModelPartHelperUtilities
+{
+
+template<class TContainerGetter>
+void AddEntitiesFromIds(
+    TContainerGetter&& rContainerGetter,
+    ModelPart* pModelPart,
+    const std::vector<IndexType>& rEntityIds)
+{
+    KRATOS_TRY
+
+    using container_type = std::remove_pointer_t<decltype(std::declval<TContainerGetter>()(std::declval<ModelPart*>()))>;
+
+    if(pModelPart->IsSubModelPart()) { //does nothing if we are on the top model part
+        //obtain from the root model part the corresponding list of nodes
+        ModelPart* root_model_part = &pModelPart->GetRootModelPart();
+
+        // we first sort and unique the given entity ids.
+        std::vector<IndexType> unique_sorted_ids;
+        unique_sorted_ids.resize(rEntityIds.size());
+        IndexPartition<IndexType>(rEntityIds.size()).for_each([&unique_sorted_ids,  &rEntityIds](const auto Index) {
+            unique_sorted_ids[Index] = rEntityIds[Index];
+        });
+        std::sort(unique_sorted_ids.begin(), unique_sorted_ids.end());
+        auto new_last = std::unique(unique_sorted_ids.begin(), unique_sorted_ids.end());
+
+        const auto& r_container = *rContainerGetter(root_model_part);
```

Comment:
ok i see you pass it as a lambda ... not sure of the why though...

Reply:
I am not sure I am understanding you correctly. I need the template argument as a functor because I want to use the same `AddEntitiesFromIds` method for all nodes/conditions/elements/master_slave_constraints/geometries. But the problem is the geometries container is in the model part, and all the rest in the mesh. Hence I could not template this function with the appropriate `TContainerTyp`, hence have to use a lambda function.

Did I answer your question?

Reply:
you did